sql/src/main/java/io/crate/planner/consumer/GroupingSubselectConsumer.java

```diff
@@ -0,0 +1,292 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.planner.consumer;
+
+import io.crate.analyze.*;
+import io.crate.analyze.relations.AnalyzedRelation;
+import io.crate.analyze.symbol.AggregateMode;
+import io.crate.analyze.symbol.Symbol;
+import io.crate.metadata.RowGranularity;
+import io.crate.planner.Limits;
+import io.crate.planner.Merge;
+import io.crate.planner.Plan;
+import io.crate.planner.PositionalOrderBy;
+import io.crate.planner.distribution.DistributionInfo;
+import io.crate.planner.node.ExecutionPhases;
+import io.crate.planner.node.dql.MergePhase;
+import io.crate.planner.projection.FilterProjection;
+import io.crate.planner.projection.GroupProjection;
+import io.crate.planner.projection.Projection;
+import io.crate.planner.projection.builder.ProjectionBuilder;
+import io.crate.planner.projection.builder.SplitPoints;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+
+public class GroupingSubselectConsumer implements Consumer {
+
+    private final Visitor visitor;
+
+    GroupingSubselectConsumer(ProjectionBuilder projectionBuilder) {
+        visitor = new Visitor(projectionBuilder);
+    }
+
+    @Override
+    public Plan consume(AnalyzedRelation relation, ConsumerContext context) {
+        return visitor.process(relation, context);
+    }
+
+    private static class Visitor extends RelationPlanningVisitor {
+
+        private final ProjectionBuilder projectionBuilder;
+
+        public Visitor(ProjectionBuilder projectionBuilder) {
+            this.projectionBuilder = projectionBuilder;
+        }
+
+        @Override
+        public Plan visitQueriedSelectRelation(QueriedSelectRelation relation, ConsumerContext context) {
+            QuerySpec querySpec = relation.querySpec();
+
+            // Check if group by is present - if not skip and continue with next consumer in chain.
```

Comment:
I'd remove the comment, this pattern should be well-known - or we should rather extend the docs in the Consumer interface or in the ConsumingPlanner.

Another option would be to add class javadocs to this class which describes which kind of statement it handles and what kind of plan it generates